### PR TITLE
Implement username availability checker

### PR DESCRIPTION
**Pull request by lukebarnard1**
Outlined here: https://github.com/vector-im/riot-web/issues/3605#issuecomment-298679388

```HTTP
GET /_matrix/.../register/available
{
    "username": "desiredlocalpart123"
}
```

If available, the response looks like
```HTTP
HTTP/1.1 200 OK
{
    "available": true
}
```

Otherwise,
```HTTP
HTTP/1.1 429
{
    "errcode": "M_LIMIT_EXCEEDED",
    "error": "Too Many Requests",
    "retry_after_ms": 2000
}
```
or
```HTTP
HTTP/1.1 400
{
    "errcode": "M_USER_IN_USE",
    "error": "User ID already taken."
}

```
or
```HTTP
HTTP/1.1 400
{
    "errcode": "M_INVALID_USERNAME",
    "error": "Some reason for username being invalid"
}
```

### CODE REVIEW

File: synapse/rest/client/v2_alpha/register.py [hunk 1]
```diff
@@ -115,6 +116,40 @@ def on_POST(self, request):
...
+        Args:
+            hs (synapse.server.HomeServer): server
+        """
+        super(UsernameAvailabilityRestServlet, self).__init__()
+        self.hs = hs
+        self.registration_handler = hs.get_handlers().registration_handler
+        self.ratelimiter = FederationRateLimiter(
+            hs.get_clock(),
+            window_size=2000, # Time window of 2s
+            sleep_limit=1, # Artificially delay requests if rate > sleep_limit/window_size
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/register.py]: `True` here surely?

File: synapse/rest/client/v2_alpha/register.py [hunk 2]
```diff
@@ -115,6 +116,40 @@ def on_POST(self, request):
...
+            sleep_msec=1000, # Amount of artificial delay to apply
+            reject_limit=1, # Error with 429 if more than reject_limit requests are queued
+            concurrent_requests=1, # Allow 1 request at a time
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/register.py]: `True`

**Comment by erikjohnston** [comment]
> $ flake8 synapse tests 
synapse/rest/client/v2_alpha/register.py:132:30: E261 at least two spaces before inline comment
synapse/rest/client/v2_alpha/register.py:133:27: E261 at least two spaces before inline comment
synapse/rest/client/v2_alpha/register.py:134:29: E261 at least two spaces before inline comment
synapse/rest/client/v2_alpha/register.py:135:28: E261 at least two spaces before inline comment
synapse/rest/client/v2_alpha/register.py:136:35: E261 at least two spaces before inline comment

File: synapse/rest/client/v2_alpha/register.py (hunk 1 above)
Comment (lukebarnard1) [inline synapse/rest/client/v2_alpha/register.py]: Er, the docs say,
`sleep_limit (int): The number of requests received `

**Pull request merged**